### PR TITLE
2608 tts training posttrain

### PR DESCRIPTION
**Pull request by shehzeen**
* Speeds reward computation by storing temporary audio in /dev/shm with automatic cleanup, while adding step-0 validation and richer rollout logging. Saves ~20 seconds per GRPO step.

* Enables predicted-phoneme GRPO: sampled phonemes are reused in teacher forcing and optimized jointly with audio using a configurable loss weight.

* Matches rollout and teacher-forced likelihood transforms - temperature based logits are used for sampling in roll-outs and also for calculating likelihood during teacher forcing.

### CODE REVIEW

File: examples/tts/easy_magpietts.py [hunk 1]
```diff
@@ -70,6 +70,9 @@ def main(cfg):
 
     model.maybe_init_from_pretrained_checkpoint(cfg=cfg)
 
+    if mode == 'onlinepo_train' and trainer.ckpt_path is None:
+        trainer.validate(model)
+
```
Comment (blisc) [inline examples/tts/easy_magpietts.py:75]: Why do you want to do validation before training?

File: nemo/collections/tts/models/easy_magpietts_preference_optimization.py [hunk 2]
```diff
@@ -348,8 +421,66 @@ def repeat_items_in_batch(self, batch: Dict, num_repeats: int) -> Dict:
...
         """Return (and create if needed) the directory used to store intermediate waveforms during PO."""
+        if not self.cfg.get('save_reward_audios', True):
+            if not hasattr(self, '_tmp_reward_audio_dir'):
+                job_id = os.environ.get('SLURM_JOB_ID', 'local')
+                self._tmp_reward_audio_dir = os.path.join(
+                    '/dev/shm', f'easy_magpie_rewards_{job_id}_rank{self.global_rank}'
+                )
+                os.makedirs(self._tmp_reward_audio_dir, exist_ok=True)
+                atexit.register(shutil.rmtree, self._tmp_reward_audio_dir, ignore_errors=True)
+            return self._tmp_reward_audio_dir
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts_preference_optimization.py:482]: Can you ensure that your code cleans up files after they are done in the `generate_and_reward` function? Or whenever it is appropriate to remove the current iteration's audio artifacts.

atexit only triggers if your programs completes without exception. And if you don't clean it up, your solution will continue to use unnecessary system memory. If you're inside a docker, it should clean up after you exit the docker, but if you run it without some form of isolation, using system memory is more delicate.

**Review by blisc — CHANGES_REQUESTED** [review]

File: examples/tts/easy_magpietts.py (hunk 1 above)
Comment (blisc) [reply examples/tts/easy_magpietts.py:75]: To get validation numbers on step 0
